analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -165,26 +173,34 @@ private void analyze(ActivityPayload payload, Set<ActivityCategoryDto> matchingA
 		}
 	}
 
-	private void analyzeInsideLock(UserAnonymizedEntityHolder userAnonymizedHolder, ActivityPayload payload,
-			Set<ActivityCategoryDto> matchingActivityCategories)
+	private void analyzeInsideLock(UserAnonymizedEntityHolder userAnonymizedHolder, ActivityPayload[] payloads)
 	{
-		updateLastMonitoredActivityDateIfRelevant(userAnonymizedHolder, payload);
-
-		Set<GoalDto> goals = determineRelevantGoals(payload, matchingActivityCategories);
-		for (GoalDto goal : goals)
+		for (ActivityPayload payload : payloads)
 		{
-			addOrUpdateActivity(userAnonymizedHolder, payload, goal);
+			updateLastMonitoredActivityDateIfRelevant(userAnonymizedHolder, payloads);
+
+			Set<GoalDto> goals = determineRelevantGoals(payload);
+			for (GoalDto goal : goals)
+			{
+				addOrUpdateActivity(userAnonymizedHolder, payload, goal);
+			}
 		}
 	}
 
 	private void updateLastMonitoredActivityDateIfRelevant(UserAnonymizedEntityHolder userAnonymizedHolder,
-			ActivityPayload payload)
+			ActivityPayload[] payloads)
 	{
-		Optional<LocalDate> lastMonitoredActivityDate = payload.userAnonymized.getLastMonitoredActivityDate();
-		LocalDate activityEndTime = payload.endTime.toLocalDate();
-		if (lastMonitoredActivityDate.map(d -> d.isBefore(activityEndTime)).orElse(true))
+		if (payloads.length == 0)
```

Comment:
Is this realistic? If it is, I'd rather not call ``get`` below, but just return when nothing is present in the ``Optional``.

Reply:
Not realistic, but theoretical. Refactored as suggested.